pyvrp/cpp/search/LocalSearch.cpp

```diff
@@ -277,8 +277,36 @@ void LocalSearch::applyOptionalClientMoves(Route::Node *U,
         update(route, route);
     }
 
-    if (!U->route())
+    if (uData.required && !U->route())
+    {
         insert(U, costEvaluator, uData.required);
+        return;
+    }
+
```

Comment:
It's assumed that U is not in a route after this, but that's not the case. So the next loop will try to insert U's that are already in the solution.